File: src/main/java/me/santio/uhc/listeners/ChatListener.java
package me.santio.uhc.listeners;

import me.santio.uhc.Game;
import me.santio.uhc.UHC;
import me.santio.uhc.models.GamePlayer;
import net.md_5.bungee.api.ChatColor;
import net.md_5.bungee.api.chat.BaseComponent;
import net.md_5.bungee.api.chat.HoverEvent;
import net.md_5.bungee.api.chat.TextComponent;
import org.bukkit.Bukkit;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.player.AsyncPlayerChatEvent;

public class ChatListener implements Listener {
    
    @SuppressWarnings("deprecation")
    @EventHandler(priority = EventPriority.HIGHEST)
    public void onChat(AsyncPlayerChatEvent event) {
        if (!event.isCancelled()) {
            GamePlayer gamePlayer = Game.getPlayers().get(event.getPlayer().getUniqueId());
            if (gamePlayer == null) return;
    
            HoverEvent hoverEvent = new HoverEvent(HoverEvent.Action.SHOW_TEXT, gamePlayer.getChatHover());
            BaseComponent[] component = TextComponent.fromLegacyText(event.getFormat());
            component[0].setHoverEvent(hoverEvent);
            
            event.setCancelled(true);
            Bukkit.getOnlinePlayers().forEach(player -> player.spigot().sendMessage(component));
        }
    }
    
}

File: src/main/kotlin/me/santio/uhc/listeners/ChatListener.kt
package me.santio.uhc.listeners

import me.santio.uhc.Game
import me.santio.uhc.UHC
import net.md_5.bungee.api.chat.HoverEvent
import net.md_5.bungee.api.chat.TextComponent
import net.md_5.bungee.api.chat.hover.content.Text
import org.bukkit.Bukkit
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.player.AsyncPlayerChatEvent
import java.util.*

object ChatListener : Listener {
    @EventHandler(priority = EventPriority.HIGHEST)
    fun onChat(event: AsyncPlayerChatEvent) {
        if (!event.isCancelled) {
            val gamePlayer = Game.players[event.player.uniqueId] ?: return

            val original: String = try { String.format(event.format, event.player.name, event.message) }
            catch(e: MissingFormatArgumentException) { event.format }

            val component = TextComponent.fromLegacyText(original)
            component[0].hoverEvent = HoverEvent(HoverEvent.Action.SHOW_TEXT, Text(gamePlayer.chatHover))

            event.isCancelled = true
            Bukkit.getOnlinePlayers().forEach { player: Player -> player.spigot().sendMessage(*component) }
        }
    }
}